### ISSUE TITLE
Pass `GraphData` to functions instead of individual attributes

### ISSUE THREAD

**Issue by francescofarina** [mlx-graphs/mlx-graphs#28]
We're currently passing individual arguments (`edge_index`, `node_features`,...) to functions and modules, which is in line with what other libraries also do.

I'm wondering whether it'd make more sense to directly pass a `GraphData` instead of the individual attributes.

**Comment by TristanBilot** [comment]
Hmm, we have to be sure that this would not add more constraints to the framework.
Ex: if we use `GraphData` in `propagate()`, we have to wrap the embeddings in a new `GraphData`.

```python
def forward(data):
  h = self.linear(data.node_embeddings)
  updated_data = mxg.graph(data.edge_index, h)
  self.propagate(updated_data)
```

Maybe for the `forward()` call it could make sense, we need to be sure it will not be problematic in some situations.

**Comment by TristanBilot** [comment]
But it could be nice as we don't have to check the shapes on inputs anymore in the message-passing part, as it would be already done by the `GraphData` constructor.

**Comment by TristanBilot** [comment]
There is another case where it can be suboptimal: some GNNs will accept as input to `forward()` additional args like `edge_weights`, `edge_features`, and others won't. How the user can know whether a given GNN model can work with edge features then? We usually check the signature of `forward()` to see if the model deals with such features, but as the edge features are wrapped in the `GraphData`, we don't need anymore to put `edge_features` in the `forward()` args and it's complicated for the user to know if he can use edge features.

I took the example of edge features but it can be the same for other types of custom features.

**Comment by francescofarina** [comment]
On the last point we can easily let the user add arbitrary attributes to the `GraphData` but which one are actually used by each GNN model can be specified in the documentation for that model instead of looking at the signature of `forward()`.

**Comment by TristanBilot** [comment]
We can do that, but do you really think it can be advantageous to use only GraphData? I don't see any real improvement yet

**Comment by francescofarina** [comment]
No, it's just a different UX that may be more convenient in some cases. I'm not saying we should go down that road but it's one worth considering. Perhaps we can wait to have a few more architectures implemented and then experiment with the different UX.

For example, by doing that we wouldn't have to always document what the `edge_index`, `node_features`,... are and the shapes we expect them to have.

**Comment by TristanBilot** [comment]
Maybe we can do something like:

```python
def forward(graph: GraphData):
  """
  ...

  Args:
    graph (GraphData[edge_index, node_features, Optional[edge_features]]): Graph with optional edge features
  """
```

**Comment by francescofarina** [comment]
Or 

```python
def forward(graph: GraphData):
  """
  ...

  Args:
    graph (GraphData): Graph with the following required attributes
      - `edge_index`
      - `node_features`
      - `edge_features` (Optional)
  """
  ```

Though we'll end up having to implement validators for each function that raise errors if the required attributes are not there. There are both pros and cons

**Comment by TristanBilot** [comment]
Yours is better!
But yes, we have to check every time...
